core/lib/spree/event/test_interface.rb

```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+require 'spree/event'
+
+module Spree
+  module Event
+    # Test helpers for {Spree::Event}
+    #
+    # This module defines test helpers methods for {Spree::Event}. They can be
+    # made available to {Spree::Event} when {Spree::Event.enable_test_interface}
+    # is called.
+    #
+    # If you prefer, you can directly call them from
+    # `Spree::Event::TestInterface}.
+    module TestInterface
+      # @see {Spree::Event::TestInterface}
+      module Methods
+        # Perform only given listeners for the duration of the block
+        #
+        # Temporarily deactivate all subscribed listeners and listen only to the
+        # provided ones for the duration of the block.
+        #
+        # @example
+        #   Spree::Event.enable_test_interface
+        #
+        #   listener1 = Spree::Event.subscribe('foo') { do_something }
+        #   listener2 = Spree::Event.subscribe('foo') { do_something_else }
+        #
+        #   Spree::Event.performing_only(listener1) do
+        #     Spree::Event.fire('foo') # This will run only `listener1`
+        #   end
+        #
+        #   Spree::Event.fire('foo') # This will run both `listener1` & `listener2`
+        #
+        # {Spree::Event::Subscriber} modules can also be given to unsubscribe from
+        # all listeners generated from it:
+        #
+        # @example
+        #   Spree::Event.performing_only(EmailSubscriber) {}
+        #
+        # You can gain more fine-grained control thanks to
+        # {Spree::Event::Subscribe#listeners}:
+        #
+        # @example
+        #   Spree::Event.performing_only(EmailSubscriber.listeners('order_finalized')) {}
+        #
+        # You can mix different ways of specifying listeners without problems:
+        #
+        # @example
+        #   Spree::Event.performing_only(EmailSubscriber, listener1) {}
+        #
+        # @param listeners_and_subscribers [Spree::Event::Listener,
+        # Array<Spree::Event::Listener>, Spree::Event::Subscriber]
+        # @yield While the block executes only provided listeners will run
+        def performing_only(*listeners_and_subscribers)
+          old_adapter = Spree::Event.default_adapter
```

Comment:
I don't understand why we need to store the old adapter, can you help me?

Reply:
I believe this is necessary because `Spree::Event.fire` uses the `default_adapter` unless one is explicitly passed in the options [here](https://github.com/solidusio/solidus/blob/master/core/lib/spree/event.rb#L73), so we have to replace that with one built by the `with_listeners` call for the duration of the tests, and then reset to the adapter instance which is not configured with a filter on the listeners.

Reply:
Exactly. The event bus is global. We change it for the duration of the `performing_only` block.

Reply:
Got it, I was confused because I wrongly interpreted `old_adapter` as `legacy_adapter`. What about changing it to `adapter_in_use` or something similar, instead?

Reply:
Oh, I see. Yeah, it can be confusing. I renamed it to `adapter_in_use` now :slightly_smiling_face: